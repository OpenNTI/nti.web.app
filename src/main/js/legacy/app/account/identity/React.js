const Ext = require('@nti/extjs');
const { getService } = require('@nti/web-client');
const { Identity } = require('@nti/web-profiles');
const LoginActions = require('internal/legacy/login/Actions');
require('internal/legacy/overrides/ReactHarness');

const AccountActions = require('../Actions');

module.exports = exports = Ext.define(
	'NextThought.app.account.identity.React',
	{
		extend: 'NextThought.ReactHarness',
		alias: 'widget.identity-react',
		component: Identity.Session,

		setTheme(scope, knockout) {},

		initComponent: function () {
			this.callParent(arguments);
			Object.assign(this.initialConfig, {
				addHistory: true,
				baseroute: '/app',
			});
			this.AccountActions = AccountActions.create();
			this.LoginActions = LoginActions.create();
		},

		async afterRender() {
			this.callParent(arguments);
			const service = await getService();
			const supportLinks = service.getSupportLinks();

			this.mon(Ext.getBody(), {
				click: this.onBodyClicks.bind(this, supportLinks),
			});
		},

		onBodyClicks(supportLinks, e) {
			let action = e.getTarget('a[data-action]')?.dataset.action;
			const link = e.getTarget('a[href]');
			const href = link?.getAttribute('href');

			// This is used to launch support logic the same way everywhere
			// we have a link that goes to the supportContact url. (so we can
			// just emit that url in the dom and not worry about coping logic)
			if (
				supportLinks.supportContact === href &&
				supportLinks.internalSupport
			) {
				action = action || 'support';
			}

			if (action) {
				if (call(this, `do-${action}`, link, e) !== false) {
					e.stopEvent();
				}
			}
		},

		showLink(target) {
			target.setAttribute('target', '_blank');
			return false;
		},

		'do-welcomeGuide'(target) {
			this.AccountActions.showWelcomePage(target.href);
		},

		'do-about': 'showLink',

		'do-privacy'(target) {
			this.AccountActions.showPrivacy(target);
		},

		'do-privacyForMinors'(target) {
			this.AccountActions.showChildrensPrivacy(target.href);
		},

		'do-terms'(target) {
			this.AccountActions.showTermsOfService(target.href);
		},

		'do-support'(target) {
			this.AccountActions.showContactUs();
		},

		'do-helpSite': 'showLink',

		'do-impersonate'() {
			this.LoginActions.handleImpersonate();
		},

		'do-logout'() {
			this.LoginActions.handleLogout();
		},
	}
);

function call(scope, action, ...args) {
	if (typeof action === 'string') {
		return call(scope, scope[action], ...args);
	}

	return action.apply(scope, args);
}
