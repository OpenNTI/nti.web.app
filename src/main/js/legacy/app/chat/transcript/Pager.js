const Ext = require('@nti/extjs');

const Globals = require('legacy/util/Globals');
const TranscriptSummary = require('legacy/model/TranscriptSummary');
const Transcript = require('legacy/model/Transcript');
const PageItem = require('legacy/store/PageItem');

const ChatActions = require('../Actions');
const ChatStateStore = require('../StateStore');

require('../components/log/PagerEntry');

module.exports = exports = Ext.define('NextThought.app.chat.transcript.Pager', {
	mixins: {
		observable: 'Ext.util.Observable',
	},

	PAGE_SIZE: 3,

	constructor: function (cfg) {
		this.initConfig(cfg || {});
		this.ChatActions = ChatActions.create();
		this.ChatStore = ChatStateStore.getInstance();

		this.mixins.observable.constructor.apply(this);
	},

	buildTranscriptStore: function (occupants) {
		var user = Ext.Array.remove(occupants.slice(), $AppConfig.username)[0],
			url = Service.getContainerUrl(
				Globals.CONTENT_ROOT,
				Globals.RECURSIVE_USER_GENERATED_DATA
			),
			s = PageItem.make(url, Globals.CONTENT_ROOT, true);

		if (!user) {
			return;
		}

		s.pageSize = this.PAGE_SIZE;
		s.proxy.extraParams = Ext.apply(s.proxy.extraParams || {}, {
			sortOn: 'CreatedTime',
			sortOrder: 'descending',
			batchSize: this.PAGE_SIZE,
			batchStart: 0,
			transcriptUser: user || '',
			accept: [
				TranscriptSummary.prototype.mimeType,
				Transcript.prototype.mimeType,
			].join(','),
		});

		this.transcriptStore = s;
		s.on('load', this.pageLoaded.bind(this));
		s.load();
	},

	bindWindow: function (win) {
		this.chatWindow = win;
		this.pages = [];
	},

	pageLoaded: function (store, records) {
		var win = this.chatWindow;

		win.maskWindow();
		this.loadTranscripts(records)
			.then(this.insertBulkMessages.bind(this))
			.then(this.maybeAddMoreEntry.bind(this, records))
			.then(win.unmaskWindow.bind(win))
			.catch(win.unmaskWindow.bind(win));
	},

	loadTranscripts: function (records) {
		var toAdd = [],
			p,
			me = this;
		Ext.each(records || [], function (record) {
			p = me.getMessages(record);
			toAdd.push(p);
		});

		return new Promise(function (fulfill, reject) {
			Promise.all(toAdd).then(function (results) {
				var allMessages = [];
				Ext.each(results, function (msgs) {
					allMessages = allMessages.concat(msgs);
				});

				fulfill(allMessages);
			});
		});
	},

	getMessages: function (transcript) {
		var me = this;

		return me.ChatActions.loadTranscript(transcript.get('RoomInfo')).then(
			function (t) {
				var messages = (t && t.get('Messages')) || [];
				return Promise.resolve(messages);
			}
		);
	},

	sortMessages: function (messages) {
		function timeSort(a, b) {
			var aRaw = a.raw || { CreatedTime: 0 },
				bRaw = b.raw || { CreatedTime: 0 };

			return aRaw.CreatedTime - bRaw.CreatedTime;
		}

		messages = Ext.Array.sort(messages, timeSort);
		return messages;
	},

	addBulkMessages: function (records) {
		this.clearPaging();

		records = this.sortMessages(records);
		if (this.chatWindow) {
			this.chatWindow.addBulkMessages(records);
		}
		return Promise.resolve();
	},

	insertBulkMessages: function (records) {
		this.clearPaging();

		records = this.sortMessages(records);
		if (this.chatWindow) {
			this.chatWindow.insertBulkMessages(0, records);
		}
		return Promise.resolve();
	},

	clearPaging: function () {
		var logView = this.chatWindow && this.chatWindow.logView;
		if (this.pagingCmp && logView) {
			logView.remove(this.pagingCmp);
			delete this.pagingCmp;
		}
	},

	maybeAddMoreEntry: function (lastLoadRecords) {
		var pageCount = Math.ceil(
				this.transcriptStore.getTotalCount() / this.PAGE_SIZE
			),
			hasMore =
				pageCount > 0
					? this.transcriptStore.currentPage < pageCount
					: false,
			logView = this.chatWindow && this.chatWindow.logView,
			me = this;

		if (logView && hasMore) {
			this.pagingCmp = logView.insert(0, {
				xtype: 'chat-pager-entry',
			});

			this.pagingCmp.onceRendered.then(function () {
				me.mon(me.pagingCmp.messageEl, 'click', me.loadNext.bind(me));
			});
		}

		return Promise.resolve();
	},

	loadNext: function () {
		var currentPage = this.transcriptStore.currentPage,
			nextPage = currentPage + 1,
			param = {
				batchStart: (nextPage - 1) * this.PAGE_SIZE,
			},
			s = this.transcriptStore;

		s.proxy.extraParams = Ext.apply(s.proxy.extraParams || {}, param);
		s.loadPage(nextPage);
	},
});
