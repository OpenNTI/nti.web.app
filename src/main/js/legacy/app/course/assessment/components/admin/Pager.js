const Ext = require('@nti/extjs');


module.exports = exports = Ext.define('NextThought.app.course.assessment.components.admin.Pager', {
	extend: 'Ext.Component',
	alias: 'widget.course-assessment-admin-pager',

	MAX_VISIBLE: 20,
	MIDDLE_RANGE: 10,

	height: 22,

	cls: 'pager',

	renderTpl: Ext.DomHelper.markup({
		cls: 'pager-container',
		cn: [
			// {tag: 'tpl', 'if': 'hasRandomAccess', cn: [
			//	{tag: 'label', cls: 'page', cn: [
			//		'Page:',
			//		{tag: 'input', type: 'text', value: '{current}'}
			//	]}
			// ]},
			{tag: 'ul', cls: 'pages', cn: [
				{tag: 'tpl', 'for': 'pages', cn: [
					{tag: 'tpl', 'if': 'isEllipse', cn: [
						{tag: 'li', cls: 'page ellipse', 'data-start': '{start}', 'data-end': '{end}', html: '{html}'}
					]},
					{tag: 'tpl', 'if': '!isEllipse', cn: [
						{tag: 'li', cls: 'page {cls}', 'data-index': '{index}', html: '{html}'}
					]}
				]}
			]}
		]
	}),


	renderSelectors: {
		pageInput: 'label input[type=text]'
	},


	afterRender: function () {
		this.callParent(arguments);

		this.mon(this.el, 'click', 'onItemClick');

		this.attachListeners();
	},


	bindStore: function (store) {
		this.store = store;

		this.mon(store, 'load', 'onStoreLoad');

		this.onStoreLoad();
	},


	getPages: function (count, current) {
		var pages = [],
			halfRange = Math.floor(this.MIDDLE_RANGE / 2);

		function addRange (start, end) {
			var i;

			for (i = start; i <= end; i++) {
				pages.push({
					html: i,
					isEllipse: false,
					index: i,
					cls: i === current ? 'active' : 'not-active'
				});
			}
		}

		function addEllipsis (start, end) {
			pages.push({
				html: '...',
				isEllipse: true,
				start: start,
				end: end
			});
		}

		// if we have less pages than the max add them all
		if (count <= this.MAX_VISIBLE) {
			addRange(1, count);
		//if the first half of the middle range is at or before the start
		} else if (current - halfRange <= 1) {
			addRange(1, this.MIDDLE_RANGE);
			addEllipsis(this.MIDDLE_RANGE + 1, count - 4);
			addRange(count - 2, count);
		//if the last half of the middle range is at or after the end
		} else if (current + halfRange >= count) {
			addRange(1, 3);
			addEllipsis(4, count - this.MIDDLE_RANGE - 1);
			addRange(count - this.MIDDLE_RANGE, count);
		} else {
			addRange(1, 3);
			addEllipsis(4, current - halfRange - 1);
			addRange(current - halfRange, current + halfRange);
			addEllipsis(current + halfRange + 1, count);
			addRange(count - 2, count);
		}

		return pages;
	},


	addNavigationPages: function (pages, pageCount, current) {
		pages.unshift({
			html: '<',
			index: 'prev',
			isEllipse: false,
			cls: current > 1 ? 'enabled' : 'disabled'
		});

		pages.push({
			html: '>',
			index: 'next',
			isEllipse: false,
			cls: current < pageCount ? 'enabled' : 'disabled'
		});
		return pages;
	},


	onStoreLoad: function () {
		var pages,
			pageCount = this.store.getTotalPages(),
			current = parseInt(this.store.getCurrentPage(), 10);

		pages = this.getPages(pageCount, current);
		pages = this.addNavigationPages(pages, pageCount, current);

		this.renderData = Ext.apply(this.renderData || {}, {
			current: current,
			pages: pages,
			hasRandomAccess: pageCount > this.MAX_VISIBLE
		});

		if (current > pageCount && pageCount > 0) {
			this.store.loadPage(pageCount);
		}

		if (this.rendered) {
			this.renderTpl.overwrite(this.el, this.renderData);
			this.attachListeners();
		}

		this.fireEvent('unmask-view');
	},


	attachListeners: function () {
		if (!this.rendered) { return; }

		var pageInput = this.el.down('label.page input[type=text]');

		Ext.destroy(this.pageInputMons);

		if (pageInput) {
			this.pageInputMons = this.mon(pageInput, {
				'keypress': 'onKeyPress',
				'destroyable': true
			});
		}
	},


	onItemClick: function (e) {
		if (e.getTarget('.disabled')) { return; }

		var page = e.getTarget('.page'),
			index = page && page.getAttribute('data-index');


		if (index === 'prev') {
			this.loadPrev();
		} else if (index === 'next') {
			this.loadNext();
		}
		else if (index === 'first') {
			this.loadFirstPage();
		}
		else if (index === 'last') {
			this.loadLastPage();
		}
		else if (index) {
			this.loadPage(index);
		}
	},


	loadPage: function (page) {
		this.fireEvent('mask-view');
		this.store.loadPage(page);
	},


	loadPrev: function () {
		var current = parseInt(this.store.getCurrentPage(), 10);

		if (current > 1) {
			this.loadPage(current - 1);
		}
	},


	loadNext: function () {
		var current = parseInt(this.store.getCurrentPage(), 10),
			total = this.store.getTotalPages();

		if (current < total) {
			this.loadPage(current + 1);
		}
	},


	loadLastPage: function () {
		var current = parseInt(this.store.getCurrentPage(), 10),
			lastPage = this.store.getTotalPages();

		if (current !== lastPage) {
			this.loadPage(lastPage);
		}
	},


	loadFirstPage: function () {
		var current = parseInt(this.store.getCurrentPage(), 10);

		if (current !== 1) {
			this.loadPage(1);
		}
	},


	getPageInputValue: function () {
		var pageInput = this.el.down('label.page input[type=text'),
			value = (pageInput && pageInput.getValue()) || 1;

		return parseInt(value, 10);
	},


	onKeyPress: function (e) {
		var keyCode = e.keyCode,
			allowed = [e.BACKSPACE, e.TAB, e.ESCAPE];

		if (keyCode === e.ENTER) {
			this.loadPage(this.getPageInputValue());
		}

		if (Ext.Array.contains(allowed, keyCode)) {
			return;
		}

		if (keyCode === 65 && e.ctrlKey) {
			return;
		}

		if ((e.shiftKey || (e.keyCode < 48 || e.keyCode > 57)) && (e.keyCode < 96 || e.keyCode > 105)) {
			e.preventDefault();
		}
	}
});
