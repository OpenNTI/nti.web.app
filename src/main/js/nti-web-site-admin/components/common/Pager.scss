@import '~@nti/style-common';

.site-admin-table-pager {
	display: flex;
	padding-top: 2.5rem;
	justify-content: center;
	margin: 0 auto;

	.page {
		cursor: pointer;
		height: 40px;
		width: 40px;
		line-height: 40px;
		text-align: center;
		border: 1px solid #e2e2e2;
		color: $secondary-grey;
		margin: 0 5px;

		&.selected {
			background-color: $primary-blue;
			color: white;
			border: 1px solid $primary-blue;
		}
	}
}
