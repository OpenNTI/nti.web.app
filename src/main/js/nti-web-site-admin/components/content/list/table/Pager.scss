@import '~@nti/style-common';

.content-table-pager {
	display: flex;
	padding-top: 2.5rem;
	width: fit-content;
	margin: 0 auto;

	.page {
		cursor: pointer;
		height: 40px;
		width: 40px;
		line-height: 40px;
		text-align: center;
		border: 1px solid #e2e2e2;
		color: $secondary-grey;
		margin: 0 5px;

		&.selected {
			background-color: $primary-blue;
			color: white;
			border: 1px solid $primary-blue;
		}
	}
}
