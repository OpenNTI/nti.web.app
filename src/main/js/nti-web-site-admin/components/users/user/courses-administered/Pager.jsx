import { DiscretePages } from '@nti/web-core';

import { Store } from './Store';

export function Pager() {
	const { totalPages, currentPage, loadPage } = Store.useProperties();

	return totalPages <= 1 ? null : (
		<DiscretePages
			mv="xl"
			total={totalPages}
			selected={currentPage}
			load={loadPage}
		/>
	);
}
