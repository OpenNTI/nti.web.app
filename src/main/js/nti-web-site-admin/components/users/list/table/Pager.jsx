import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

export default class Pager extends React.Component {
	static propTypes = {
		store: PropTypes.object.isRequired,
		pageNumber: PropTypes.number,
		numPages: PropTypes.number
	}

	renderPage = (num) => {
		const {pageNumber, store} = this.props;

		const cls = cx('page', { selected: pageNumber === num + 1 });

		return (
			<div
				key={num}
				onClick={()=>{
					if(store.loadPage) {
						store.loadPage(num + 1);
					}
					else {
						store.goToPage(num + 1);
					}
				}}
				className={cls}>
				{num + 1}
			</div>
		);
	}

	render () {
		const {numPages, pageNumber} = this.props;

		const startPage = pageNumber - 4 < 0 ? 0 : pageNumber - 4;
		const lastPage = pageNumber + 2 >= numPages ? numPages - 1 : pageNumber + 2;

		let numbers = [];

		for(let i = startPage; i <= lastPage; i++) {
			numbers.push(i);
		}

		if(numbers.length <= 1) {
			return null;
		}

		return (
			<div className="users-table-pager">
				{(numbers).map(this.renderPage)}
			</div>
		);
	}
}
