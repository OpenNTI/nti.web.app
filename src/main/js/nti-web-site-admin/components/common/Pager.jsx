import './Pager.scss';
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import {Connectors} from '@nti/lib-store';

export default
@Connectors.Any.connect({
	loadPage: 'loadPage'
})
class Pager extends React.Component {
	static propTypes = {
		loadPage: PropTypes.func.isRequired,
		pageNumber: PropTypes.number,
		numPages: PropTypes.number
	}

	renderPage = (num) => {
		const {pageNumber, loadPage} = this.props;

		const cls = cx('page', { selected: pageNumber === num + 1 });

		return (
			<div
				key={num}
				onClick={()=>{
					loadPage(num + 1);
				}}
				className={cls}>
				{num + 1}
			</div>
		);
	}

	render () {
		const {numPages, pageNumber} = this.props;

		const startPage = pageNumber - 4 < 0 ? 0 : pageNumber - 4;
		const lastPage = pageNumber + 2 >= numPages ? numPages - 1 : pageNumber + 2;

		let numbers = [];

		for(let i = startPage; i <= lastPage; i++) {
			numbers.push(i);
		}

		if(numbers.length <= 1) {
			return null;
		}

		return (
			<div className="site-admin-table-pager">
				{(numbers).map(this.renderPage)}
			</div>
		);
	}
}
