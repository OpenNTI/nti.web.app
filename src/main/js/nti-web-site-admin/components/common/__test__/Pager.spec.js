/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';

import Pager from '../Pager';

describe('Site admin table pager test', () => {
	test('Test no pages', () => {
		const pagerCmp = renderer.create(<Pager loadPage={()=>{}} numPages={0} />);

		const tree = pagerCmp.toJSON();

		// 0 pages should result in nothing being rendered
		expect(tree).toMatchSnapshot();
	});

	test('Test 1 page', () => {
		const pagerCmp = renderer.create(<Pager loadPage={()=>{}} numPages={1} pageNumber={1} />);

		const tree = pagerCmp.toJSON();

		// 1 page should also result in nothing being rendered since there is only one page option
		expect(tree).toMatchSnapshot();
	});

	test('Test 2 pages', () => {
		const pagerCmp = renderer.create(<Pager loadPage={()=>{}} numPages={2} pageNumber={1} />);

		const tree = pagerCmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Test 10 pages, pageNumber=1', () => {
		const pagerCmp = renderer.create(<Pager loadPage={()=>{}} numPages={10} pageNumber={1} />);

		const tree = pagerCmp.toJSON();

		// will show the first pages available
		expect(tree).toMatchSnapshot();
	});

	test('Test 10 pages, pageNumber=5', () => {
		const pagerCmp = renderer.create(<Pager loadPage={()=>{}} numPages={10} pageNumber={5} />);

		const tree = pagerCmp.toJSON();

		// will show several pages back and several pages forward
		expect(tree).toMatchSnapshot();
	});
});
