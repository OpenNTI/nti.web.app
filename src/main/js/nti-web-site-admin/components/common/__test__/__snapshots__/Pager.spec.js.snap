// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site admin table pager test Test 1 page 1`] = `null`;

exports[`Site admin table pager test Test 2 pages 1`] = `
<div
  className="site-admin-table-pager"
>
  <div
    className="page selected"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    2
  </div>
</div>
`;

exports[`Site admin table pager test Test 10 pages, pageNumber=1 1`] = `
<div
  className="site-admin-table-pager"
>
  <div
    className="page selected"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    2
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    3
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    4
  </div>
</div>
`;

exports[`Site admin table pager test Test 10 pages, pageNumber=5 1`] = `
<div
  className="site-admin-table-pager"
>
  <div
    className="page"
    onClick={[Function]}
  >
    2
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    3
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    4
  </div>
  <div
    className="page selected"
    onClick={[Function]}
  >
    5
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    6
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    7
  </div>
  <div
    className="page"
    onClick={[Function]}
  >
    8
  </div>
</div>
`;

exports[`Site admin table pager test Test no pages 1`] = `null`;
